### PR TITLE
Implement scan_ipc

### PR DESCRIPTION
**Pull request by Sicheng-Pan**
This PR aims to implement `scan_ipc`, which reads data from IPC file or stream. It also includes a minimal unit test for this functionality. For Rust codes, changes in `src/rust/src/rdataframe` are related to the implementation, while some other unrelated files are modified by `cargo fmt` as well.

### CODE REVIEW

**Comment by sorhawell** [comment]
this looks very good :)  I will some remarks on Monday when I'm back at the office.

File: R/ipc.R [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#' new LazyFrame from ipc file
```
Comment (eitsupi) [inline R/ipc.R:1]:
```suggestion
#' new LazyFrame from Apache Arrow IPC file
```

File: src/rust/src/conversion_r_to_s.rs [hunk 2]
```diff
@@ -1,20 +1,19 @@
+use extendr_api::prelude::*;
 /// this file implements any conversion from Robject to polars::Series
 /// most other R to polars conversion uses the module only pub function robjname2series()
-
 use polars::prelude as pl;
-use extendr_api::prelude::*;
-use polars::prelude::NamedFrom;
 use polars::prelude::IntoSeries;
+use polars::prelude::NamedFrom;
```
Comment (eitsupi) [inline src/rust/src/conversion_r_to_s.rs:6]:
```suggestion
use polars::prelude::{IntoSeries, NamedFrom};
```

File: tests/testthat/test-ipc.R [hunk 3]
```diff
@@ -0,0 +1,9 @@
+test_that("Test reading data from ipc", {
+  tmpf = tempfile()
```
Comment (eitsupi) [inline tests/testthat/test-ipc.R:2]:
```suggestion
  tmpf = tempfile()
  on.exit(unlink(tmpf))
```

**Review by eitsupi — COMMENTED** [review]
This is awesome!

However, in this context, I think "ipc" needs to be called "Apache Arrow IPC file".
Please refer to the following mailing list.
https://lists.apache.org/thread/hph6ckgmvp18y3o5gy3s9x7th9wq7kyq

File: src/rust/src/conversion_r_to_s.rs (hunk 2 above)
Comment (Sicheng-Pan) [reply src/rust/src/conversion_r_to_s.rs]: Fixed.

File: R/ipc.R (hunk 1 above)
Comment (Sicheng-Pan) [reply R/ipc.R]: Updated.

File: tests/testthat/test-ipc.R (hunk 3 above)
Comment (Sicheng-Pan) [reply tests/testthat/test-ipc.R:2]: Fixed.

I also notice some other unit tests using `tempfile()` missing `on.exit(unlink(...))`. Should I update them as well?

Comment (eitsupi) [reply tests/testthat/test-ipc.R:2]: > I also notice some other unit tests using `tempfile()` missing `on.exit(unlink(...))`. Should I update them as well?

Thanks for finding it.
If you are willing to take on that fix, another PR would be fine. (We can merge this PR without waiting for that fix)

Comment (Sicheng-Pan) [reply tests/testthat/test-ipc.R:2]: I'll make another PR for the fix

**Comment by Sicheng-Pan** [comment]
@sorhawell May I ask if there is anything else I should do for this PR?

**Review by sorhawell — CHANGES_REQUESTED** [review]
@Sicheng-Pan sorry my mistake. I made some comments, but forgot to press the "submit review" button. I thought you were able to see them already.

File: src/rust/src/rdataframe/read_ipc.rs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+#[extendr]
+pub fn new_from_ipc(
+    path: String,
+    n_rows: Nullable<i64>,
+    cache: bool,
+    rechunk: bool,
+    row_name: Nullable<String>,
+    row_count: i64,
+    memmap: bool,
+) -> List {
```
[the remark this answers is not in the document]
Comment (Sicheng-Pan) [reply src/rust/src/rdataframe/read_ipc.rs]: Fixed

File: src/rust/src/rdataframe/read_ipc.rs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    let args = ScanArgsIpc {
+        n_rows: null_to_opt(n_rows).and_then(|n| usize::try_from(n).ok()),
```
[the remark this answers is not in the document]
Comment (Sicheng-Pan) [reply src/rust/src/rdataframe/read_ipc.rs]: Fixed.

Should we update other files (e.g. `read_parquet.rs`) for proper conversion handling?

File: R/ipc.R [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    path,
+    n_rows,
+    cache,
+    rechunk,
+    row_count_name,
+    row_count_offset,
+    memmap
+  )
+
+  unwrap(result_lf)
```
[the remark this answers is not in the document]
Comment (Sicheng-Pan) [reply R/ipc.R]: Fixed

File: tests/testthat/test-ipc.R [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+  on.exit(unlink(tmpf))
+  arrow::write_ipc_file(iris, tmpf)
```
[the remark this answers is not in the document]
Comment (Sicheng-Pan) [reply tests/testthat/test-ipc.R:6]: Partially fixed. I'm a bit not sure how to test if some arguments are used as expected (e.g. `cache`)

[the remark this answers is not in the document]
Comment (sorhawell) [reply tests/testthat/test-ipc.R:6]: true, just skip cache then :)

File: tests/testthat/test-ipc.R [hunk 8]
```diff
@@ -0,0 +1,24 @@
+test_that("Test reading data from Apache Arrow IPC", {
+
+  # Put data in Apache Arrow IPC format
+  tmpf = tempfile()
+  on.exit(unlink(tmpf))
```
Comment (sorhawell) [inline tests/testthat/test-ipc.R:5]: you can add in as first line  `skip_if_not_installed("arrow")`then test will not fail soft dependency arrow is not installed

File: src/rust/src/rdataframe/read_ipc.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sorhawell) [reply src/rust/src/rdataframe/read_ipc.rs]: nice looks good

File: src/rust/src/rdataframe/read_ipc.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (sorhawell) [reply src/rust/src/rdataframe/read_ipc.rs]: @Sicheng-Pan yes that was some of the first I wrote a year ago, I could barely code rust back then :)

[the remark this answers is not in the document]
Comment (sorhawell) [reply src/rust/src/rdataframe/read_ipc.rs]: in another pr...

**Comment by sorhawell** [comment]
looks good, just a few more tests for what makes sense and it is a wrap

File: tests/testthat/test-ipc.R (hunk 8 above)
Comment (Sicheng-Pan) [reply tests/testthat/test-ipc.R:8]: Fixed

File: tests/testthat/test-ipc.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (Sicheng-Pan) [reply tests/testthat/test-ipc.R:9]: Fixed

**Review by sorhawell — APPROVED** [review]

**Pull request merged**